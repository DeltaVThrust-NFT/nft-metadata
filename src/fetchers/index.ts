import { fetchUnderlyingContractData } from './defaultFetcher'
import { getAddress } from '@ethersproject/address'

export function fetcherLookup(contractAddress: string) {
  const safeAddress = getAddress(contractAddress)
  switch (safeAddress) {
    default:
      return fetchUnderlyingContractData
  }
}
