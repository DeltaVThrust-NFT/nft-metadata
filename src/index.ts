export * from './constants'
export * from './metadata'
export * from './agent'
export * from './uri'
