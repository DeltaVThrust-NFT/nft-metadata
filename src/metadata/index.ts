export * from './fetcher'
export * from './hashmasks'
export * from './loot'
export * from './normaliser'
export * from './zora'
