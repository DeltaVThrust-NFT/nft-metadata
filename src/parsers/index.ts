import { getAddress } from '@ethersproject/address'
import { parseGenericMetadata } from './openseaMetadataParser'
import {
  ART_BLOCKS_CURATED_TOKEN_ADDRESS,
  ART_BLOCKS_TOKEN_ADDRESS,
  HASHMASKS_TOKEN_ADDRESS,
  MAKERSPLACE_TOKEN_ADDRESS,
  ZORA_TOKEN_ADDRESS,
} from '../constants'
import { parseZoraMetadata } from './zoraMetadataParser'
import { parseMakersplaceMetadata } from './makersplaceMetadataParser'
import { parseHashmasksMetadata } from './hashmasksMetadataParser'
import { parseArtblocksMetadata } from './artblocksMetadataParser'

export function parserLookup(contractAddress: string) {
  const safeAddress = getAddress(contractAddress)
  switch (safeAddress) {
    case ART_BLOCKS_TOKEN_ADDRESS:
      return parseArtblocksMetadata
    case ART_BLOCKS_CURATED_TOKEN_ADDRESS:
      return parseArtblocksMetadata
    case HASHMASKS_TOKEN_ADDRESS:
      return parseHashmasksMetadata
    case MAKERSPLACE_TOKEN_ADDRESS:
      return parseMakersplaceMetadata
    case ZORA_TOKEN_ADDRESS:
      return parseZoraMetadata
    default:
      return parseGenericMetadata
  }
}
