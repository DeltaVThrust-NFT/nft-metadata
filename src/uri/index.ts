export { getAlternateContractCall } from './contract'
export {
  fetchMimeType,
  fetchURI,
  fetchWithRetries,
  fetchWithTimeout,
  parseDataUri,
} from './fetch'
export { isIPFS, getIPFSUrl, getPrivateGateway } from './ipfs'
export { getStaticURI, getURIData } from './static'
