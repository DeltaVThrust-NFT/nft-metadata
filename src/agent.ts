import { JsonRpcProvider } from '@ethersproject/providers'
import { fetchOwnerOf, getSupportedInterfaces } from './utils/contracts'
import { Erc721Factory } from '@zoralabs/core/dist/typechain'
import { Fetcher, fetcherLookup, FetcherResponse } from './fetchers'
import { fetchMetadata, fetchMimeType } from './utils/fetch'
import { getIPFSUrl } from './utils/ipfs'
import { Parser, parserLookup } from './parsers'

export interface TokenPKDto {
  networkId: number
  tokenAddress: string
  tokenId: string
}

export interface NftMetadata extends TokenPKDto {
  metadata: any

  name?: string
  description?: string

  ownerAddress: string
  creatorAddress?: string

  tokenURI: string
  tokenURL: string
  tokenURLMimeType: string

  contentURL?: string
  contentURLMimeType?: string

  imageURL?: string
  imageURLMimeType?: string

  externalLink?: string

  attributes?: Record<string, any>[]
}

export enum Network {
  MAINNET = 1,
  RINKEBY = 4,
}

export type AgentNetworkConfiguration = {
  [key: number]: JsonRpcProvider
}

export type AgentConfiguration = {
  ipfsGateway?: string
  fetchTimeout?: number
  providers: AgentNetworkConfiguration
  parsers?: {
    [key: string]: Parser
  }
  fetchers?: {
    [key: string]: Fetcher
  }
}

export type TokenContractData = FetcherResponse & TokenPKDto & Record<any, any>
export type TokenData = { metadata: Record<any, any>; contentType: string }

export class Agent {
  constructor(private readonly options: AgentConfiguration) {}

  private getProviderForNetwork(number: Network): JsonRpcProvider {
    const provider = this.options.providers?.[number]
    if (!provider) {
      throw new Error('No provider for network')
    }
    return provider
  }

  public async supportedInterfaces(networkId: Network, tokenAddress: string) {
    const provider = this.getProviderForNetwork(networkId)
    return getSupportedInterfaces(provider, tokenAddress)
  }

  public async fetchContractData(
    networkId: Network,
    tokenAddress: string,
    tokenId: string,
  ) {
    const provider = this.getProviderForNetwork(networkId)
    const { erc721, erc721Metadata, erc721Enumerable } =
      await this.supportedInterfaces(networkId, tokenAddress)

    if (!erc721) {
      throw new Error(
        `${tokenAddress} on network: ${networkId} does not support ERC721 standard`,
      )
    }

    const erc721Contract = Erc721Factory.connect(tokenAddress, provider)
    const ownerAddress = await fetchOwnerOf(
      erc721Contract,
      tokenId,
      // Can catch if contract can make a total supply call
      // to handle open zeppelin burn
      erc721Enumerable,
    )

    let tokenURI: string | undefined
    if (erc721Metadata) {
      tokenURI = await erc721Contract.tokenURI(tokenId)
    }

    const baseData = {
      networkId,
      tokenAddress,
      tokenId,
      tokenURI,
      ownerAddress,
    }

    // bespoke fetches for other underlying data
    const fetcher = fetcherLookup(tokenAddress, this.options.fetchers)
    if (fetcher) {
      const data = await fetcher({
        provider,
        tokenAddress,
        tokenId,
      })

      return {
        ...baseData,
        ...data,
      }
    }

    return baseData
  }

  public async parseMetadata(
    contractData: TokenContractData,
    tokenData: TokenData,
  ) {
    const metadata = tokenData.metadata

    let meta: Partial<NftMetadata> = {}

    if (contractData.tokenURI) {
      meta.tokenURL = getIPFSUrl(
        contractData.tokenURI,
        this.options.ipfsGateway,
      )
      meta.tokenURLMimeType = tokenData.contentType
    }

    if (metadata.image) {
      meta.imageURL = getIPFSUrl(metadata.image, this.options.ipfsGateway)
      meta.contentURL = meta.imageURL
    }

    if (metadata.animation_url) {
      meta.contentURL = getIPFSUrl(
        metadata.animation_url,
        this.options.ipfsGateway,
      )
    }

    const {
      name,
      description,
      attributes,
      traits,
      external_url: externalURL,
      mimeType,
    } = metadata

    if (meta.contentURL) {
      meta.contentURLMimeType = await fetchMimeType(
        meta.contentURL,
        {
          timeout: this.options.fetchTimeout,
        },
        mimeType,
      )
    }

    if (meta.imageURL && meta.imageURL !== meta.contentURL) {
      meta.imageURLMimeType = await fetchMimeType(meta.imageURL, {
        timeout: this.options.fetchTimeout,
      })
    } else {
      delete meta.imageURL
    }

    const baseMeta = {
      metadata,
      ...meta,
      ...(name && { name }),
      ...(description && { description }),
      ...(attributes && { attributes }),
      ...(traits && { attributes: traits }),
      ...(externalURL && { externalURL }),
    }

    // Return any extra fields unparseable by defaults
    // @TODO - @ethandaya move this behaviour to a dynamic registry
    const parser = parserLookup(contractData.tokenAddress, this.options.parsers)
    if (parser) {
      const meta = await parser({
        provider: this.getProviderForNetwork(contractData.networkId),
        contractData,
        tokenData,
        ipfsGateway: this.options.ipfsGateway,
        fetchTimeout: this.options.fetchTimeout,
        baseMeta,
      })
      return {
        ...baseMeta,
        ...meta,
      }
    }

    return baseMeta
  }

  public async fetchAndParseTokenData(
    networkId: number,
    tokenAddress: string,
    tokenId: string,
  ) {
    const contractData = await this.fetchContractData(
      networkId,
      tokenAddress,
      tokenId,
    )

    if (!contractData.tokenURI) {
      return { ...contractData }
    }

    const tokenData = await fetchMetadata(contractData.tokenURI, {
      timeout: this.options.fetchTimeout,
      ipfsGateway: this.options.ipfsGateway,
    })

    const metadata = await this.parseMetadata(contractData, tokenData)

    return {
      ...contractData,
      ...metadata,
    }
  }
}
