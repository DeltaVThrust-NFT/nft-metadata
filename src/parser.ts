import { JsonRpcProvider } from '@ethersproject/providers'
import { getAddress } from '@ethersproject/address'
import { fetcherLookup } from './fetchers'
import { parserLookup } from './parsers'

export interface NftMetadata {
  metadata: any

  name?: string
  description?: string

  ownerAddress: string

  tokenURI: string
  contentURI: string

  externalLink?: string
  attributes?: Record<string, any>[]
}

export class Parser {
  constructor(private readonly provider: JsonRpcProvider) {}

  public async fetchUnderlyingContractData(
    contractAddress: string,
    tokenId: string,
  ) {
    const fetcher = fetcherLookup(contractAddress)
    return fetcher(this.provider, contractAddress, tokenId)
  }

  public async parseTokenMetadata(
    contractAddress: string,
    tokenId: string,
    tokenURI: string,
  ) {
    const parser = parserLookup(contractAddress)
    return parser(this.provider, contractAddress, tokenId, tokenURI)
  }

  public async fetchAndParseTokenMeta(
    tokenAddress: string,
    tokenId: string,
  ): Promise<NftMetadata> {
    const safeAddress = getAddress(tokenAddress)

    const { tokenURI, ownerAddress } = await this.fetchUnderlyingContractData(
      safeAddress,
      tokenId,
    )

    const parsedMetadata = await this.parseTokenMetadata(
      safeAddress,
      tokenId,
      tokenURI,
    )

    return {
      tokenURI,
      ownerAddress,
      ...parsedMetadata,
    }
  }
}
