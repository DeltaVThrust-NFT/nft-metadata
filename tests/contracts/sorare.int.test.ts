import { Agent, SORARE_TOKEN_ADDRESS } from '../../src'
import { testProvider } from '../setupProvider'
import { isAddress } from '@ethersproject/address'

const SORARE_CRITERIA = {
  input: {
    tokenId:
      '100408090256748841933639219919068439822835068705742947267506906666220789418990',
    tokenAddress: SORARE_TOKEN_ADDRESS,
    networkId: 1,
  },
}

describe('Sorare ERC721', () => {
  const parser = new Agent({
    providers: {
      1: testProvider,
    },
    ipfsGateway: 'https://ipfs.fleek.co',
    fetchTimeout: 60000,
  })

  beforeEach(() => {
    jest.setTimeout(120000)
  })

  it(`should be able to fetch and parse metadata for token id: ${SORARE_CRITERIA.input.tokenId} on network: ${SORARE_CRITERIA.input.networkId}`, async () => {
    const { ownerAddress, attributes, metadata, ...meta } =
      await parser.fetchAndParseTokenData(
        1,
        SORARE_CRITERIA.input.tokenAddress,
        SORARE_CRITERIA.input.tokenId,
      )
    expect(meta).toMatchSnapshot()
    expect(attributes).toBeDefined()
    expect(attributes).toBeDefined()
    expect(isAddress(ownerAddress)).toBeTruthy()
  })
})
