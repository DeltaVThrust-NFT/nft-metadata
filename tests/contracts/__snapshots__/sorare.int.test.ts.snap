// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sorare ERC721 should be able to fetch and parse metadata for token id: 100408090256748841933639219919068439822835068705742947267506906666220789418990 on network: 1 1`] = `
Object {
  "contentURL": "https://assets.sorare.com/card/044a4d0a-2c44-41a8-b81f-ecd69cbe031a/picture/40af00c4aca787c49865c09aec66182e.png",
  "contentURLMimeType": "image/png",
  "description": "Rare Player Cards are only issued at 100 editions per season.",
  "externalURL": "?referrer=opensea",
  "name": "Hiroyuki Abe 2021-22 â€¢ Rare 38/100",
  "networkId": 1,
  "tokenAddress": "0x629a673a8242c2ac4b7b8c5d8735fbeac21a6205",
  "tokenId": "100408090256748841933639219919068439822835068705742947267506906666220789418990",
  "tokenURI": "https://api.sorare.com/api/v1/cards/100408090256748841933639219919068439822835068705742947267506906666220789418990",
  "tokenURL": "https://api.sorare.com/api/v1/cards/100408090256748841933639219919068439822835068705742947267506906666220789418990",
  "tokenURLMimeType": "application/json; charset=utf-8",
}
`;
